pkg/controller/garbagecollector/garbage_collector.go

```diff
@@ -0,0 +1,501 @@
+		orphanQueue:                      workqueue.NewTimedWorkQueue(),
+		registeredRateLimiter:            NewRegisteredRateLimiter(deletableResources),
+		registeredRateLimiterForMonitors: NewRegisteredRateLimiter(deletableResources),
+		absentOwnerCache:                 NewUIDCache(500),
+	}
+	gc.propagator = &Propagator{
+		eventQueue: workqueue.NewTimedWorkQueue(),
+		uidToNode: &concurrentUIDToNode{
+			RWMutex:   &sync.RWMutex{},
+			uidToNode: make(map[types.UID]*node),
+		},
+		gc: gc,
+	}
+	for resource := range deletableResources {
+		if _, ok := ignoredResources[resource]; ok {
+			glog.V(6).Infof("ignore resource %#v", resource)
+			continue
+		}
+		kind, err := gc.restMapper.KindFor(resource)
+		if err != nil {
+			return nil, err
+		}
+		monitor, err := gc.monitorFor(resource, kind)
+		if err != nil {
+			return nil, err
+		}
+		gc.monitors = append(gc.monitors, monitor)
+	}
+	return gc, nil
+}
+
+func (gc *GarbageCollector) Run(workers int, stopCh <-chan struct{}) {
+	defer gc.dirtyQueue.ShutDown()
+	defer gc.orphanQueue.ShutDown()
+	defer gc.propagator.eventQueue.ShutDown()
+
+	glog.Infof("Garbage Collector: Initializing")
+	for _, monitor := range gc.monitors {
+		go monitor.controller.Run(stopCh)
+	}
+
+	var syncs []cache.InformerSynced
+	for _, monitor := range gc.monitors {
+		syncs = syncs.append(monitor.controller.HasSynced())
+	}
+	if !cache.WaitForCacheSync(stopCh, syncs...) {
+		return
+	}
+	glog.Infof("Garbage Collector: All monitored resources synced. Proceeding to collect garbage")
+
+	// worker
+	go wait.Until(gc.propagator.processEvent, 0, stopCh)
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(gc.worker, 0, stopCh)
+		go wait.Until(gc.orphanFinalizer, 0, stopCh)
+	}
+	Register()
+	<-stopCh
+	glog.Infof("Garbage Collector: Shutting down")
+}
+
+func (gc *GarbageCollector) worker() {
+	timedItem, quit := gc.dirtyQueue.Get()
+	if quit {
+		return
+	}
+	defer gc.dirtyQueue.Done(timedItem)
+	err := gc.processItem(timedItem.Object.(*node))
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("Error syncing item %#v: %v", timedItem.Object, err))
+		// retry if garbage collection of an object failed.
+		gc.dirtyQueue.Add(timedItem)
+		return
+	}
+	DirtyProcessingLatency.Observe(sinceInMicroseconds(gc.clock, timedItem.StartTime))
+}
+
+func objectReferenceToMetadataOnlyObject(ref objectReference) *metaonly.MetadataOnlyObject {
+	return &metaonly.MetadataOnlyObject{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: ref.APIVersion,
+			Kind:       ref.Kind,
+		},
+		ObjectMeta: v1.ObjectMeta{
+			Namespace: ref.Namespace,
+			UID:       ref.UID,
+			Name:      ref.Name,
+		},
+	}
+}
+
+// classify the latestReferences to three categories:
+// solid: the owner exists, and is not "waiting"
+// dangling: the owner does not exist
+// waiting: the owner exists, its deletionTimestamp is non-nil, and it has
+// FinalizerDeletingDependents
+// This function communicates with the server.
```

Comment:
We rely on the server communication to handle the cases where the watch event of the owner's creation comes later than the dependent's creation. I left TODO's saying we can avoid the communication in other cases. Let's resolve the TODO in a follow-up PR, because it might cause flakes.